README.md
# Scoped Tokens for Webex Administrators [![published](https://static.production.devnetcloud.com/codeexchange/assets/images/devnet-published.svg)](https://developer.cisco.com/codeexchange/github/repo/ObjectIsAdvantag/webex-integration-admin)

A Web application that generates scoped tokens for administrators of Webex Organizations.

The application comes in two flavors:
1. [dynamic](server.js): let's the admin choose from a list of 'spark-admin' scopes (default mode and illustrated below)
2. [static](static.js): uses an hard-coded list of admin scopes.

## Test the application

[Click to live test](https://webex-token.herokuapp.com) the dynamic OAuth tokens issuer for Webex admins.

0. Wait for up to 30 seconds for the application to launch, since the app is powered by [Heroku free dynos](https://devcenter.heroku.com/articles/free-dyno-hours)

1. Select a few scopes, and click `Start` to initiate the OAuth flow. 

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;![](docs/img/select-scopes.png)


2. If permissions are granted, the application will display newly issued access & refresh tokens to invoke Webex APIs on the user's behalf.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;![](docs/img/OAuth-Show-Tokens.png)


Note that:
- this example is configured with a pre-defined [Webex OAuth Integration](https://developer.webex.com/docs/integrations) that includes all Webex scopes (except spark:all), and redirects to http://localhost:8080/oauth so that Webex Administrators can quickly generate 'scoped' OAuth access tokens on their local machine.
- for production purpose, you would typically create your own OAuth integration from the Webex for Developers portal. You would also deploy your integration over HTTPS.


To learn more about the implementation details of this Webex OAuth integration, check the DevNet learning lab: [Run a Webex Teams Integration locally](https://developer.cisco.com/learning/tracks/collab-cloud/spark-apps/collab-spark-intl/step/1).


## Run the application locally

``` bash
git clone https://github.com/ObjectIsAdvantag/webex-integration-admin
cd webex-integration-admin
npm install
DEBUG=oauth* node server.js
```

You're all set, the integration is ready to use.

Open [http://localhost:8080](http://localhost:8080) in a Web browser.

Select scopes in the proposed list, 
and Click the 'Start' button


## Customize the scoped tokens

Use the `static.js` service to launch an application with a pre-defined list of scopes.

To define the list of scopes requested: 
- open the `.env` file and update the SCOPES varialbe, 
- or specify custom scopes on the command line

**Example to request custom compliance scopes**

Run the command below in a bash terminal:

```
DEBUG=oauth* SCOPES="spark-compliance:rooms_read spark-compliance:teams_read spark-compliance:events_read"  node static.js
```


## Registering your own OAuth Integration

To learn more about Webex Integrations, [read the reference documentation](https://developer.webex.com/authentication.html).

In the example above, we were running a pre-registered OAuth integration running on port 8080.
Let's now see how to create your own Webex integration that we will be running on port **9090** for example.

Go to the [Webex for developers](https://developer.webex.com) portal, click 'Apps', 'add Apps', and then [create an integration](https://developer.webex.com/add-integration.html).

Fill the form fields:
- **Name**: your app name, such as 'My Awesome App' in the snapshot below,
- **Description**: these details are not displayed to your app end-users,
- **Support email**: a valid email address for Webex Spark operations team to reach to you if necessary,
- **App Icon**: simply pick one of the default icons or upload a 512x512 file. Feel free to use this provided sample for now: [https://bit.ly/SparkIntegration-512x512](https://bit.ly/SparkIntegration-512x512)
- **Redirect URI**: enter `http://localhost:9090/oauth` since this is the URL at which the provided code sample will be listening on your local machine. More to come in the next steps if this lab
- **Scopes**: select all scopes but spark:all.

> Note the list of scopes you selected corresponds to the maximum set of scopes that your integration will be entitled to ask for. However, from code, your integration can dynamically [refine the set of scopes asked for](https://github.com/CiscoDevNet/webex-integration-sample/blob/master/server.js#L30) in order to comply with the real needs of your application.

A good practice is to start small and extend the set of scopes asked for as the end-users gain trust in your app and is ready to ask more advanced features.

Click save, and look for your integration `client id` and `client secret`.


Let's now configure the integration: you can either paste your integration client id and secret into [the code](https://github.com/CiscoDevNet/webex-integration-sample/blob/master/server.js#L26), or set these as env variables on the command line.

Instructions for **Mac, Linux and Windows bash users** 
- open a terminal
- on a single command line, type:

    ```shell
    DEBUG=oauth* PORT=9090 REDIRECT_URI="http://localhost:9090/oauth" CLIENT_ID="YOUR_INTEGRATION_ID" CLIENT_SECRET="YOUR_INTEGRATION_SECRET"  node server.js
    ```

Instructions for **Windows command shell users**
- open a command shell
- enter the commands below:

    ```shell
    set DEBUG="oauth*"
    set PORT=9090
    set REDIRECT_URI="http://localhost:9090/oauth"
    set CLIENT_ID="YOUR_INTEGRATION_ID"
    set CLIENT_SECRET="YOUR_INTEGRATION_SECRET"
    node server.js
    ```

**You're all set. Restart your integration, and give it a try: [http://localhost:9090](http://localhost:9090)**


This step by step guide explained how to register a Webex Integration running on a local developer environment.
In the real-world, you will want to deploy your integration over HTTPS, and register its public URL rather than your local machine's.

Check the DevNet learning lab: ['Deploy a Webex OAuth Integration'](https://developer.cisco.com/learning/lab/collab-spark-intd-heroku/step/1) for detailled instructions.


## Common error codes

The following errors may be raised as the OAuth flow completes.

### invalid_scope: 'No user entitled service in request scopes'

The Webex user can NOT access the requested scopes.

Generally happens when a standard Webex user is requesting 'admin' level scopes.


### access_denied

The Webex organisation has configured policies that does not allow this OAuth integration to be executed for the Webex user.

Contact the administrator of the Webex organisation for the user, or use a different Webex account (example: gmail accounts).


## More about Webex Integrations

Webex Integrations are a way for your apps to request permission to invoke the Webex APIs on behalf of Webex Teams users. 
The process used to request permission is called an OAuth Grant Flow, and is documented in the [Integrations guide](https://developer.webex.com/docs/integrations).

You can experiment the flow in DevNet Learning lab ["Understand the OAuth Grant flow of Webex Teams Integrations"](https://learninglabs.cisco.com/tracks/collab-cloud/business-messaging/collab-spark-auth/step/1).


CHANGELOG
# ChangeLog

## v0.6 - 20200826
- updated list of scopes and organized in 2 columns
- changed api domain to https://webexapis.com
- aligned UX for static and dynamic examples
- fix for the static example

## v0.5
- new page to display access & refresh tokens (with explanations)
- do not log tokens when running in production

## v0.4
- hosting on Heroku for quick test purposes

## v0.3
- new possibility to select admin scopes
- distinct static & dynamic services


static.js
//
// Copyright (c) 2019 Cisco Systems
// Licensed under the MIT License 
//

/*
 * a Webex Teams Integration based on Node.js, that acts on user's behalf.
 * implements the Webex OAuth flow, to retreive an API access tokens.
 * 
 * See documentation: https://developer.webex.com/authentication.html
 * 
 */

// Load environment variables from project .env file
require('node-env-file')(__dirname + '/.env');

const debug = require("debug")("oauth");
const fine = require("debug")("oauth:fine");

const request = require("request");
const express = require('express');
const app = express();


// Step 0: create an OAuth integration from https://developer.webex.com/add-integration.html
//   - then fill in your Integration properties below

// A default proposed integration has been setup with:
//   - all available Webex user, admin & compliance scopes (except: spark:all)
//   - redirect URI to localhost:8080/oauth 
//
const clientId = process.env.CLIENT_ID || "C24ba23a9215b84f92b608ac72d664a4c0416e8c6c5575f64c438813d50ba1955";
const clientSecret = process.env.CLIENT_SECRET || "34dbd055b33a3da0ad61c0ba3f0bff98a49cd90518d0042dc48e8c221184c14a";

// supported scopes are documented at: https://developer.webex.com/add-integration.html, the scopes separator is a space
const scopes = process.env.SCOPES || "spark-admin:people_read";

// Compute redirect URI where your integration is waiting for Webex cloud to redirect and send the authorization code
// unless provided via the REDIRECT_URI variable
const port = process.env.PORT || 8080;
const localRedirectURI = `http://localhost:${port}/oauth`;
let redirectURI = process.env.REDIRECT_URI
if (!redirectURI) {
   // Glitch hosting
   if (process.env.PROJECT_DOMAIN) {
      redirectURI = "https://" + process.env.PROJECT_DOMAIN + ".glitch.me/oauth";
   }
   else {
      // defaults to localhost
      redirectURI = `http://localhost:${port}/oauth`;
   }
}
debug(`OAuth integration settings:\n   - CLIENT_ID    : ${clientId}\n   - REDIRECT_URI : ${redirectURI}\n   - SCOPES       : ${scopes}`);


// Step 1: initiate the OAuth flow
//   - serves a Web page with a link to the Webex OAuth flow initializer
//
// Initiate the OAuth flow from the 'index.ejs' template  
// ------------------------------------------------------------- 
// -- Comment this section to initiate the flow from  static html page

// state can be used for security and/or correlation purposes
const state = process.env.STATE || "CiscoDevNet";

/*
 * Initiate the flow with the default scopes of the OAuth Integration
 */
const initiateURL = "https://webexapis.com/v1/authorize?"
   + "client_id=" + clientId
   + "&response_type=code"
   + "&redirect_uri=" + encodeURIComponent(redirectURI)
   + "&scope=" + encodeURIComponent(scopes)
   + "&state=" + state;

const read = require("fs").readFileSync;
const join = require("path").join;
const str = read(join(__dirname, '/www/index.ejs'), 'utf8');
const ejs = require("ejs");
const compiled = ejs.compile(str)({
   "link": initiateURL,
   "local": (redirectURI == localRedirectURI)
});

app.get("/index.html", function (req, res) {
   debug("serving the integration home page (generated from an EJS template)");
   res.send(compiled);
});

app.get("/", function (req, res) {
   res.redirect("/index.html");
});

// -------------------------------------------------------------
// Statically serve the "/www" directory
// WARNING: Do not move the 2 lines of code below, as we need this exact precedance order for the static and dynamic HTML generation to work correctly all together
//          If the section above is commented, the static index.html page will be served instead of the EJS template.
const path = require('path');
app.use("/", express.static(path.join(__dirname, 'www')));


// Step 2: process OAuth Authorization codes
//
app.get("/oauth", function (req, res) {
   debug("oauth callback hitted");

   // Did the user decline
   if (req.query.error) {
      if (req.query.error == "access_denied") {
         debug("user declined, received err: " + req.query.error);
         sendError("user declined access", res);
         return;
      }

      if (req.query.error == "invalid_scope") {
         debug("invalid scope requested, received err: " + req.query.error + ", description: " + req.query.error_description);
         sendError(req.query.error_description, res);
         return;
      }

      if (req.query.error == "server_error") {
         debug("server error, received err: " + req.query.error);
         sendError(req.query.error, res);
         return;
      }

      debug("unhandled error: " + req.query.error);
      sendError(req.query.error, res);
      return;
   }

   // Check request parameters correspond to the spec
   if ((!req.query.code) || (!req.query.state)) {
      debug("expected code & state query parameters are not present");
      sendError("expected code & state query parameters are not present", res);
      return;
   }

   // Check State 
   // [NOTE] we implement a Security check below, but the State variable can also be leveraged for Correlation purposes
   if (state != req.query.state) {
      debug("State does not match");
      sendError("unexpected state", res);
      return;
   }

   // Retreive access token (expires in 14 days) & refresh token (expires in 90 days)
   //   { 
   //      "access_token":"N2MxMmE0YzgtMjY0MS00MDIxLWFmZDItNTg0MGVkOWEyNWQ3YmMzMmFlODItYzAy",
   //      "expires_in":1209599,
   //      "refresh_token":"NjBjNDk3MjktMjUwMy00YTlkLWJkOTctM2E2MjE3YWU1NmI4Njk3Y2IzODctMjBh",
   //      "refresh_token_expires_in":7775999
   //   }
   const options = {
      method: "POST",
      url: "https://webexapis.com/v1/access_token",
      headers: {
         "content-type": "application/x-www-form-urlencoded"
      },
      form: {
         grant_type: "authorization_code",
         client_id: clientId,
         client_secret: clientSecret,
         code: req.query.code,
         redirect_uri: redirectURI
      }
   };
   request(options, function (error, response, body) {
      if (error) {
         debug("could not reach Webex cloud to retreive access & refresh tokens, err: " + error.message);
         sendError("communication error: " + error.message, res);
         return;
      }

      if (response.statusCode != 200) {
         debug("access token not issued with status code: " + response.statusCode);
         switch (response.statusCode) {
            case 400:
               const responsePayload = JSON.parse(response.body);
               debug("bad request: " + responsePayload.description);
               sendError("bad request: ", res);
               break;
            case 401:
               debug("authentication error: " + error.message);
               sendError("authentication error: " + error.message, res);
               break;
            default:
               debug("unhandled error: " + error.message);
               sendError("unhandled error: " + error.message, res);
               break;
         }
         return;
      }

      // Check payload
      const json = JSON.parse(body);
      if ((!json) || (!json.access_token) || (!json.expires_in) || (!json.refresh_token) || (!json.refresh_token_expires_in)) {
         debug("could not parse access & refresh tokens");
         sendError("could not parse response from Webex", res);
         return;
      }
      debug("OAuth flow completed, fetched tokens: " + JSON.stringify(json));

      // OAuth flow has completed
      oauthFlowCompleted(json, res);
   });
});


// Step 3: this is where the integration runs its custom logic
//   - this function is called as the OAuth flow has been successfully completed, 
//   - this function is expected to send back an HTML page to the end-user
//   
// some optional activities to perform here: 
//    - associate the issued access token to a user through the state (acting as a Correlation ID)
//    - store the refresh token (valid 90 days) to reissue later a new access token (valid 14 days)
function oauthFlowCompleted(json, res) {

   //
   // Custom logic below
   //

   const str = read(join(__dirname, '/www/show-tokens.ejs'), 'utf8');

   const compiled = ejs.compile(str)({
      "accessToken": json.access_token,
      "accessTokenExpires": Math.round(json.expires_in/(3600*24)),
      "refreshToken": json.refresh_token,
      "refreshTokenExpires": Math.round(json.refresh_token_expires_in/(3600*24)),
      "local": (redirectURI == localRedirectURI)
   });

   res.send(compiled);
}


//
// Example of Refresh token usage
//
function refreshToken(token) {

   const options = {
      method: "POST",
      url: "https://webexapis.com/v1/access_token",
      headers: {
         "content-type": "application/x-www-form-urlencoded"
      },
      form: {
         grant_type: "refresh_token",
         client_id: clientId,
         client_secret: clientSecret,
         refresh_token: refresh_token
      }
   };
   request(options, function (error, response, body) {
      if (error) {
         debug("could not reach Webex cloud to refresh access token");
         return;
      }

      if (response.statusCode != 200) {
         debug("access token not issued with status code: " + response.statusCode);
         return;
      }

      // Check payload
      const json = JSON.parse(body);
      if ((!json) || (!json.access_token) || (!json.expires_in) || (!json.refresh_token) || (!json.refresh_token_expires_in)) {
         debug("could not parse response");
         return;
      }

      // Refresh token obtained
      debug("newly issued tokens: " + JSON.stringify(json));
   });
}


function getLogoutURL(token, redirectURL) {
   const rootURL = redirectURL.substring(0, redirectURL.length - 5);
   return "https://idbroker.webex.com/idb/oauth2/v1/logout?"
      + "goto=" + encodeURIComponent(rootURL)
      + "&token=" + token;
}


function sendError(error, res) {
   const str = read(join(__dirname, '/www/error-page.ejs'), 'utf8');
   const compiled = ejs.compile(str)({ "error": error });
   res.send(compiled);
}



// Starts the Webex Integration
app.listen(port, function () {
   console.log("Webex OAuth Integration started on port: " + port);
});

server.js
//
// Copyright (c) 2019 Cisco Systems
// Licensed under the MIT License 
//

/*
 * a Webex Teams Integration based on Node.js, that acts on user's behalf.
 * implements the Webex OAuth flow, to retreive an API access tokens.
 * 
 * See documentation: https://developer.webex.com/authentication.html
 * 
 */

// Load environment variables from project .env file
require('node-env-file')(__dirname + '/.env');

const debug = require("debug")("oauth");
const fine = require("debug")("oauth:fine");

const request = require("request");
const express = require('express');
const app = express();


// Step 0: create an OAuth integration from https://developer.webex.com/add-integration.html
//   - then fill in your Integration properties below

// A default proposed integration has been setup with:
//   - all available Webex user, admin & compliance scopes (except: spark:all)
//   - redirect URI to localhost:8080/oauth 
//
const clientId = process.env.CLIENT_ID || "C24ba23a9215b84f92b608ac72d664a4c0416e8c6c5575f64c438813d50ba1955";
const clientSecret = process.env.CLIENT_SECRET || "34dbd055b33a3da0ad61c0ba3f0bff98a49cd90518d0042dc48e8c221184c14a";

// Compute redirect URI where your integration is waiting for Webex cloud to redirect and send the authorization code
// unless provided via the REDIRECT_URI variable
const port = process.env.PORT || 8080;
const localRedirectURI = `http://localhost:${port}/oauth`;
let redirectURI = process.env.REDIRECT_URI
if (!redirectURI) {
   // Glitch hosting
   if (process.env.PROJECT_DOMAIN) {
      redirectURI = "https://" + process.env.PROJECT_DOMAIN + ".glitch.me/oauth";
   }
   else {
      // defaults to localhost
      redirectURI = localRedirectURI;
   }
}
debug(`OAuth integration settings:\n   - CLIENT_ID    : ${clientId}\n   - REDIRECT_URI : ${redirectURI}\n   - SCOPES       : self-service via checkboxes`);


// Step 1: initiate the OAuth flow
//   - serves a Web page with a link to the Webex OAuth flow initializer
//
// Initiate the OAuth flow from the 'index.ejs' template  
// ------------------------------------------------------------- 
// -- Comment this section to initiate the flow from  static html page

// state can be used for security and/or correlation purposes
const state = process.env.STATE || "CiscoDevNet";

/*
 * Initiate the flow of the OAuth Integration
 */
const initiateURL = "https://webexapis.com/v1/authorize?"
   + "client_id=" + clientId
   + "&response_type=code"
   + "&redirect_uri=" + encodeURIComponent(redirectURI)
   // NOTE: commenting out the line below as the scopes are dynamically selected in this example
   //   + "&scope=" + encodeURIComponent(scopes)
   + "&state=" + state;

const read = require("fs").readFileSync;
const join = require("path").join;
const str = read(join(__dirname, '/www/scopes.ejs'), 'utf8');
const ejs = require("ejs");
const compiled = ejs.compile(str)({
   "link": initiateURL,
   "local": (redirectURI == localRedirectURI)
});

app.get("/index.html", function (req, res) {
   debug("serving the integration home page (generated from an EJS template)");
   res.send(compiled);
});

app.get("/", function (req, res) {
   res.redirect("/index.html");
});

// -------------------------------------------------------------
// Statically serve the "/www" directory
// WARNING: Do not move the 2 lines of code below, as we need this exact precedance order for the static and dynamic HTML generation to work correctly all together
//          If the section above is commented, the static index.html page will be served instead of the EJS template.
const path = require('path');
app.use("/", express.static(path.join(__dirname, 'www')));


// Step 2: process OAuth Authorization codes
//
app.get("/oauth", function (req, res) {
   debug("oauth callback hitted");

   // Did the user decline
   if (req.query.error) {
      if (req.query.error == "access_denied") {
         debug("user declined, received err: " + req.query.error);
         sendError("user declined access", res);
         return;
      }

      if (req.query.error == "invalid_scope") {
         debug("invalid scope requested, received err: " + req.query.error + ", description: " + req.query.error_description);
         sendError(req.query.error_description, res);
         return;
      }

      if (req.query.error == "server_error") {
         debug("server error, received err: " + req.query.error);
         sendError(req.query.error, res);
         return;
      }

      debug("unhandled error: " + req.query.error);
      sendError(req.query.error, res);
      return;
   }

   // Check request parameters correspond to the spec
   if ((!req.query.code) || (!req.query.state)) {
      debug("expected code & state query parameters are not present");
      sendError("expected code & state query parameters are not present", res);
      return;
   }

   // Check State 
   // [NOTE] we implement a Security check below, but the State variable can also be leveraged for Correlation purposes
   if (state != req.query.state) {
      debug("State does not match");
      sendError("unexpected state", res);
      return;
   }

   // Retreive access token (expires in 14 days) & refresh token (expires in 90 days)
   //   { 
   //      "access_token":"N2MxMmE0YzgtMjY0MS00MDIxLWFmZDItNTg0MGVkOWEyNWQ3YmMzMmFlODItYzAy",
   //      "expires_in":1209599,
   //      "refresh_token":"NjBjNDk3MjktMjUwMy00YTlkLWJkOTctM2E2MjE3YWU1NmI4Njk3Y2IzODctMjBh",
   //      "refresh_token_expires_in":7775999
   //   }
   const options = {
      method: "POST",
      url: "https://webexapis.com/v1/access_token",
      headers: {
         "content-type": "application/x-www-form-urlencoded"
      },
      form: {
         grant_type: "authorization_code",
         client_id: clientId,
         client_secret: clientSecret,
         code: req.query.code,
         redirect_uri: redirectURI
      }
   };
   request(options, function (error, response, body) {
      if (error) {
         debug("could not reach Webex cloud to retreive access & refresh tokens, err: " + error.message);
         sendError("communication error: " + error.message, res);
         return;
      }

      if (response.statusCode != 200) {
         debug("access token not issued with status code: " + response.statusCode);
         switch (response.statusCode) {
            case 400:
               const responsePayload = JSON.parse(response.body);
               debug("bad request: " + responsePayload.description);
               sendError("bad request: ", res);
               break;
            case 401:
               debug("authentication error: " + error.message);
               sendError("authentication error: " + error.message, res);
               break;
            default:
               debug("unhandled error: " + error.message);
               sendError("unhandled error: " + error.message, res);
               break;
         }
         return;
      }

      // Check payload
      const json = JSON.parse(body);
      if ((!json) || (!json.access_token) || (!json.expires_in) || (!json.refresh_token) || (!json.refresh_token_expires_in)) {
         debug("could not parse access & refresh tokens");
         sendError("could not parse response from Webex", res);
         return;
      }

      // SECURITY: log tokens if in development environmet
      if ((localRedirectURI == redirectURI) || ('development' == process.env.NODE_ENV)) {
         debug("OAuth flow completed, fetched tokens: " + JSON.stringify(json));
      }

      // OAuth flow has completed
      oauthFlowCompleted(json, res);
   });
});


// Step 3: this is where the integration runs its custom logic
//   - this function is called as the OAuth flow has been successfully completed, 
//   - this function is expected to send back an HTML page to the end-user
//   
// some optional activities to perform here: 
//    - associate the issued access token to a user through the state (acting as a Correlation ID)
//    - store the refresh token (valid 90 days) to reissue later a new access token (valid 14 days)
function oauthFlowCompleted(json, res) {

   //
   // Custom logic below
   //

   const str = read(join(__dirname, '/www/show-tokens.ejs'), 'utf8');
   const compiled = ejs.compile(str)({
      "accessToken": json.access_token,
      "accessTokenExpires": Math.round(json.expires_in / (3600 * 24)),
      "refreshToken": json.refresh_token,
      "refreshTokenExpires": Math.round(json.refresh_token_expires_in / (3600 * 24)),
      "local": (redirectURI == localRedirectURI)
   });
   res.send(compiled);
}


//
// Example of Refresh token usage
//
function refreshToken(token) {

   const options = {
      method: "POST",
      url: "https://webexapis.com/v1/access_token",
      headers: {
         "content-type": "application/x-www-form-urlencoded"
      },
      form: {
         grant_type: "refresh_token",
         client_id: clientId,
         client_secret: clientSecret,
         refresh_token: refresh_token
      }
   };
   request(options, function (error, response, body) {
      if (error) {
         debug("could not reach Webex cloud to refresh access token");
         return;
      }

      if (response.statusCode != 200) {
         debug("access token not issued with status code: " + response.statusCode);
         return;
      }

      // Check payload
      const json = JSON.parse(body);
      if ((!json) || (!json.access_token) || (!json.expires_in) || (!json.refresh_token) || (!json.refresh_token_expires_in)) {
         debug("could not parse response");
         return;
      }

      // Refresh token obtained
      debug("newly issued tokens: " + JSON.stringify(json));
   });
}


function getLogoutURL(token, redirectURL) {
   const rootURL = redirectURL.substring(0, redirectURL.length - 5);
   return "https://idbroker.webex.com/idb/oauth2/v1/logout?"
      + "goto=" + encodeURIComponent(rootURL)
      + "&token=" + token;
}


function sendError(error, res) {
   const str = read(join(__dirname, '/www/error-page.ejs'), 'utf8');
   const compiled = ejs.compile(str)({ "error": error });
   res.send(compiled);
}



// Starts the Webex Integration
app.listen(port, function () {
   console.log("Webex OAuth Integration started on port: " + port);
});